### PR TITLE
#29 - User profile backend

### LINKED ISSUE #29: User profile - Backend

### PR DESCRIPTION
**Pull request by PrabuckiDominik**
This should close #29.
Added:
 - User profile that displays user data.

### CODE REVIEW

File: resources/js/Pages/Auth/Profile.vue [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<script setup lang="ts">
+import {defineProps} from 'vue'
+import { router } from '@inertiajs/vue3'
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/Profile.vue:3]: Do we really need this router? Please look here https://github.com/blumilksoftware/keating/blob/main/resources/js/Pages/Dashboard/PasswordUpdate.vue

File: resources/js/Pages/Auth/Profile.vue [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+function submit() {
+  router.post('/profile/forgot-password')
+}
+</script>
+
+<template>
+  <div>
+    <p>Imię: {{ props.user.name }}</p>
+    <p>Nazwisko: {{ props.user.surname }}</p>
+    <p>Email: {{ props.user.email }}</p>
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/Profile.vue:27]:
```suggestion
    <p>E-mail: {{ props.user.email }}</p>
```

File: resources/js/Pages/Auth/Profile.vue [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    <p>Szkoła: {{ props.user.school.name }}</p>
+    <form @submit.prevent="submit">
+      <button type="submit">Zmiana hasła</button>
+    </form>
+    <div v-if="status" class="alert alert-success">{{ status }}</div>
+  </div>
+</template>
+
+<style scoped>
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/Profile.vue:36]: Do we need this?

File: resources/js/Pages/Auth/Profile.vue [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+</script>
+
+<template>
+  <div>
+    <p>Imię: {{ props.user.name }}</p>
+    <p>Nazwisko: {{ props.user.surname }}</p>
+    <p>Email: {{ props.user.email }}</p>
+    <p>Szkoła: {{ props.user.school.name }}</p>
+    <form @submit.prevent="submit">
+      <button type="submit">Zmiana hasła</button>
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/Profile.vue:30]: When the user is logged into application, why he receives link to reset password? I believe that it should be a form, where he fills his old password and a new one and confirmation of the new one

File: app/Http/Controllers/ProfileUserController.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+use App\Http\Requests\Auth\ForgotPasswordRequest;
+use Illuminate\Http\RedirectResponse;
+use Inertia\Inertia;
+use Inertia\Response;
+
+class ProfileUserController extends Controller
+{
+    public function create(): Response
+    {
+        $user = auth()->user()->load("school");
```
Comment (kamilpiech97) [inline app/Http/Controllers/ProfileUserController.php:16]:
```suggestion
    public function create(Request $request): Response
    {
        $user = $request->user()->load("school");
```

File: app/Http/Controllers/ProfileUserController.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+                "user" => $user,
+                "status" => session("status"),
+            ],
+        );
+    }
+
+    public function forgotPassword(): RedirectResponse
+    {
+        $userEmail = auth()->user()->email;
+        $request = ForgotPasswordRequest::create("/forgot-password", "POST", ["email" => $userEmail]);
```
Comment (kamilpiech97) [inline app/Http/Controllers/ProfileUserController.php:30]: same here

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: tests/Feature/ResetPasswordTest.php [hunk 7]
```diff
@@ -115,4 +115,65 @@ public function testUserCannotResetPasswordWithMismatchedPasswords(): void
...
+                "current_password" => "current-password",
+                "password" => "new-password",
+                "password_confirmation" => "new-password",
+            ])
+            ->assertRedirect("/profile")
+            ->assertSessionHas("success", "Zaktualizowano hasło");
+        $this->assertTrue(Hash::check("new-password", $user->fresh()->password));
+    }
+
+    public function testAuthUserCanNotResetPasswordViaProfilePasswordResetWithWrongPassword(): void
```
Comment (EwelinaSkrzypacz) [inline tests/Feature/ResetPasswordTest.php:137]:
```suggestion
    public function testAuthUserCannotResetPasswordWithWrongPassword(): void
```

File: tests/Feature/ResetPasswordTest.php [hunk 8]
```diff
@@ -115,4 +115,65 @@ public function testUserCannotResetPasswordWithMismatchedPasswords(): void
...
+                "current_password" => "wrong-current-password",
+                "password" => "new-password",
+                "password_confirmation" => "new-password",
+            ])
+            ->assertRedirect("/profile")
+            ->assertSessionHasErrors(["current_password" => trans("validation.current_password", ["attribute" => "current password"])]);
+        $this->assertTrue(Hash::check("current-password", $user->fresh()->password));
+    }
+
+    public function testAuthUserCanNotResetPasswordViaProfilePasswordResetWithNotMatchingNewPasswords(): void
```
Comment (EwelinaSkrzypacz) [inline tests/Feature/ResetPasswordTest.php:155]:
```suggestion
    public function testAuthUserCannotResetPasswordWithNotMatchingNewPasswords(): void
```

File: tests/Feature/ResetPasswordTest.php [hunk 9]
```diff
@@ -115,4 +115,65 @@ public function testUserCannotResetPasswordWithMismatchedPasswords(): void
...
+                "current_password" => "current-password",
+                "password" => "new-password",
+                "password_confirmation" => "different-password",
+            ])
+            ->assertRedirect("/profile")
+            ->assertSessionHasErrors(["password" => trans("validation.confirmed", ["attribute" => "hasło"])]);
+        $this->assertTrue(Hash::check("current-password", $user->fresh()->password));
+    }
+
+    public function testUserCanNotAccessResetPasswordViaProfilePasswordReset(): void
```
Comment (EwelinaSkrzypacz) [inline tests/Feature/ResetPasswordTest.php:173]:
```suggestion
    public function testUserCannotAccessResetPassword(): void
```

File: tests/Feature/ResetPasswordTest.php [hunk 10]
```diff
@@ -115,4 +115,65 @@ public function testUserCannotResetPasswordWithMismatchedPasswords(): void
 
         $this->assertTrue(Hash::check("oldPassword", $user->fresh()->password));
     }
+
+    public function testAuthUserCanResetPasswordViaProfilePasswordReset(): void
```
Comment (EwelinaSkrzypacz) [inline tests/Feature/ResetPasswordTest.php:119]:
```suggestion
    public function testAuthUserCanResetPassword(): void
```

File: routes/web.php [hunk 11]
```diff
@@ -19,6 +20,7 @@
 
 Route::get("/", [ContestController::class, "index"])->name("home");
 Route::post("/auth/register", [RegisterUserController::class, "store"])->name("register");
+Route::get("/auth/login", fn() => redirect("/"))->name("login");
```
Comment (EwelinaSkrzypacz) [inline routes/web.php:23]: The route below has the same name

File: resources/js/Pages/Auth/PasswordUpdate.vue [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    },
+  })
+}
+
+</script>
+
+<template>
+  <div>
+    <form @submit.prevent="updatePassword">
+      <label for="current_password">Current Password:</label>
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/PasswordUpdate.vue:30]: Why in English?

File: routes/web.php [hunk 13]
```diff
@@ -28,9 +30,17 @@
...
     Route::get("/auth/forgot-password", [PasswordResetLinkController::class, "create"])->name("password.request");
     Route::post("/auth/forgot-password", [PasswordResetLinkController::class, "store"])->name("password.email");
-    Route::get("/auth/password/reset/{token}", [PasswordResetLinkController::class, "resetCreate"])->name("password.reset");
 });
 
+Route::middleware("auth")->group(function (): void {
+    Route::get("/auth/logout", [AuthenticateSessionController::class, "logout"])->name("logout");
+    Route::get("/profile", [ProfileUserController::class, "create"])->name("profile");
+    Route::get("/profile/password-edit", [ProfileUserController::class, "edit"])->name("profile.password.edit");
+    Route::post("/profile/forgot-password", [ProfileUserController::class, "forgot"])->name("profile.password.forgot");
```
Comment (kamilpiech97) [inline routes/web.php:39]: Is it needed for logged user?

File: app/Http/Controllers/ProfileUserController.php [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+    public function forgot(Request $request): RedirectResponse
+    {
+        $userEmail = $request->user()->email;
+        $request = ForgotPasswordRequest::create("/forgot-password", "POST", ["email" => $userEmail]);
+        $passwordResetLinkController = new PasswordResetLinkController();
+        $passwordResetLinkController->store($request);
+
+        return redirect()->route("profile")->with("status", session("status"));
+    }
```
Comment (kamilpiech97) [inline app/Http/Controllers/ProfileUserController.php:44]: Is it for logged user?

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: resources/js/Pages/Auth/PasswordUpdate.vue [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+</script>
+
+<template>
+  <div>
+    <form @submit.prevent="updatePassword">
+      <label for="current_password">Aktualne Hasło:</label>
+      <input v-model="form.current_password" name="password" type="password" required><br>
+      <div v-if="errors.current_password">{{ props.errors.current_password }}</div>
+      <label for="password">Nowe Hasło:</label>
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/PasswordUpdate.vue:33]:
```suggestion
      <label for="password">Nowe hasło:</label>
```

File: resources/js/Pages/Auth/PasswordUpdate.vue [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+      <input v-model="form.password" name="password" type="password"><br>
+      <div v-if="errors.password">{{ props.errors.password }}</div>
+      <label for="password_confirmation">Potwierdzenie Nowego Hasła:</label>
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/PasswordUpdate.vue:36]:
```suggestion
      <label for="password_confirmation">Potwierdzenie nowego hasła:</label>
```

File: resources/js/Pages/Auth/PasswordUpdate.vue [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+      <input v-model="form.password_confirmation" name="password_confirmation" type="password"><br>
+      <div v-if="errors.password_confirmation">{{ props.errors.password_confirmation }}</div>
+      <button type="submit">Utwórz</button>
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/PasswordUpdate.vue:39]:
```suggestion
        <button type="submit">Zmień hasło</button>
```

File: app/Http/Controllers/ProfileUserController.php [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+    public function edit(Request $request): Response
+    {
+        return Inertia::render("Auth/PasswordUpdate");
+    }
+
+    public function update(Request $request, Hasher $hasher): RedirectResponse
+    {
+        $validated = $request->validate([
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/ProfileUserController.php:37]: I would suggest to validate through form request https://laravel.com/docs/11.x/validation#form-request-validation

File: resources/js/Pages/Auth/PasswordUpdate.vue [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+      <input v-model="form.current_password" name="password" type="password" required><br>
+      <div v-if="errors.current_password">{{ props.errors.current_password }}</div>
+      <label for="password">Nowe hasło:</label>
+      <input v-model="form.password" name="password" type="password"><br>
+      <div v-if="errors.password">{{ props.errors.password }}</div>
+      <label for="password_confirmation">Potwierdzenie nowego hasła:</label>
+      <input v-model="form.password_confirmation" name="password_confirmation" type="password"><br>
+      <div v-if="errors.password_confirmation">{{ props.errors.password_confirmation }}</div>
+        <button type="submit">Zmień hasło</button>
+    </form>
```
Comment (AleksandraKozubal) [inline resources/js/Pages/Auth/PasswordUpdate.vue:40]: Resolve GitHub Actions warning from above

Comment (PrabuckiDominik) [reply resources/js/Pages/Auth/PasswordUpdate.vue:40]: Fixed.

File: routes/web.php (hunk 13 above)
Comment (PrabuckiDominik) [reply routes/web.php]: Fixed.

File: app/Http/Controllers/ProfileUserController.php (hunk 14 above)
Comment (PrabuckiDominik) [reply app/Http/Controllers/ProfileUserController.php]: Fixed.

File: resources/js/Pages/Auth/Profile.vue (hunk 2 above)
Comment (PrabuckiDominik) [reply resources/js/Pages/Auth/Profile.vue]: Fixed.

**Review by kamilpiech97 — APPROVED** [review]

**Review by EwelinaSkrzypacz — APPROVED** [review]

**Pull request merged**